### PR TITLE
Open-source CockroachDB ranges check

### PR DESCRIPTION
**Pull request by timaa2k**
## High-level description

CockroachDB is introduced to DC/OS in version 1.13. To make sure that no master node replacements/upgrades can happen too quickly for CockroachDB nodes to rebalance their data we introduce also this dcos-check from DC/OS Enterprise.

This PR adds the dcos-check for CockroachDB underreplicated ranges to DC/OS from DC/OS Enterprise. It also adds a test that replaces all master nodes in 3 master cluster one after another to test that the check works correctly.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS_OSS-4753](https://jira.mesosphere.com/browse/DCOS_OSS-4753) Add dcos-check fo cockroachdb replication to DC/OS.

The plan is to have a follow-up PR for Enterprise that incorporates this check from dcos-checks.

## Checklist for all PRs

  - [ ] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change: This is considered part of the CockroachDB open-sourcing effort.
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___

### CODE REVIEW

**Comment by timaa2k** [comment]
@mesosphere-mergebot label Ready For Review

File: test-e2e/test_master_node_replacement.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+            dcos_config=dcos_config,
+            ip_detect_path=ip_detect_path,
+        )
+        wait_for_dcos_oss(
+            cluster=original_cluster,
+            request=request,
+            log_dir=log_dir,
+        )
+        current_cluster = original_cluster
+        tmp_clusters = set([])
```
Comment (jongiddy) [inline test-e2e/test_master_node_replacement.py]: Minor point: `set()` creates an empty set without first creating an empty list.

File: test-e2e/test_master_node_replacement.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+        original_masters = original_cluster.masters
+
+        try:
+            for master_to_be_replaced in original_masters:
+                # Destroy a master and free one IP address.
+                original_cluster.destroy_node(node=master_to_be_replaced)
```
Comment (jongiddy) [inline test-e2e/test_master_node_replacement.py]: You could use `current_cluster` here, right? I think that would be clearer, as, conceptually, the original cluster no longer exists (at least after the first iteration).

File: test-e2e/test_master_node_replacement.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                (new_master, ) = temporary_cluster.masters
+                new_master.install_dcos_from_path(
+                    dcos_installer=artifact_path,
+                    dcos_config=dcos_config,
+                    role=Role.MASTER,
+                    ip_detect_path=ip_detect_path,
+                )
+                # Form a new cluster with the newly create master node.
+                new_cluster = Cluster.from_nodes(
+                    masters=current_cluster.masters.union(set([new_master])),
```
Comment (jongiddy) [inline test-e2e/test_master_node_replacement.py]: `current_cluster.masters.add(new_master)` should give you the same thing.

File: test-e2e/test_master_node_replacement.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                    public_agents=current_cluster.public_agents,
+                )
+                # The `wait_for_dcos_oss` function waits until the new master has
+                # joined the cluster and all masters are healthy. Without the
+                # cockroachdb check, this succeeds before all cockroachdb ranges
+                # have finished replicating to the new master. That meant that the
+                # next master would be replaced too quickly, while it had data that
+                # was not present elsewhere in the cluster. This lead to
+                # irrecoverable dataloss.  This function waits until the node is
+                # "healthy". It is assumed that this performs the same diagnostics
```
Comment (jongiddy) [inline test-e2e/test_master_node_replacement.py]: Can we be more definite here and remove the assumption? "This performs the same diagnostics..."

File: test-e2e/test_master_node_replacement.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                # we instruct operators to perform when we tell them to "wait until
+                # the cluster becomes healthy before replacing the next master
+                # node."
+                #
+                # We don't call the cockroachdb ranges check directly as the
+                # purpose of this test is to ensure that when an operator follows
+                # our documented procedure for replacing a master node multiple
+                # times in a row then his cluster remains healthy throughout and
```
Comment (jongiddy) [inline test-e2e/test_master_node_replacement.py]: `his` -> `the`

File: test-e2e/test_master_node_replacement.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+                # was not present elsewhere in the cluster. This lead to
+                # irrecoverable dataloss.  This function waits until the node is
+                # "healthy". It is assumed that this performs the same diagnostics
+                # we instruct operators to perform when we tell them to "wait until
+                # the cluster becomes healthy before replacing the next master
+                # node."
+                #
+                # We don't call the cockroachdb ranges check directly as the
+                # purpose of this test is to ensure that when an operator follows
+                # our documented procedure for replacing a master node multiple
```
Comment (jongiddy) [inline test-e2e/test_master_node_replacement.py]: Can we link to the documented procedure?

Comment (timaa2k) [reply test-e2e/test_master_node_replacement.py]: See at the end of the comment section.

**Comment by timaa2k** [comment]
PTAL @jongiddy @ahairape @mhrabovcin

**Comment by timaa2k** [comment]
Test result of running the tests added here on CI:
https://teamcity.mesosphere.io/viewLog.html?buildId=1588869&
I don't know why it is not yet automatically executed with the PR.

**Pull request merged**